src/main/java/org/opensearch/index/codec/customcodecs/Lucene99QatCodec.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.codec.customcodecs;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.codecs.FilterCodec;
+import org.apache.lucene.codecs.StoredFieldsFormat;
+import org.apache.lucene.codecs.lucene99.Lucene99Codec;
+import org.opensearch.index.codec.PerFieldMappingPostingFormatCodec;
+import org.opensearch.index.mapper.MapperService;
+
+import java.util.Set;
+import java.util.function.Supplier;
+
+import com.intel.qat.QatZipper;
+
+/**
+ * Extends {@link FilterCodec} to reuse the functionality of Lucene Codec. Supports two modes zstd
+ * and zstd_no_dict. Uses Lucene99 as the delegate codec
+ *
+ * @opensearch.internal
+ */
+public abstract class Lucene99QatCodec extends FilterCodec {
+
+    /** Default compression level used for compression */
+    public static final int DEFAULT_COMPRESSION_LEVEL = 3;
+
+    /** The default QAT mode */
+    public static final QatZipper.Mode DEFAULT_QAT_MODE = QatZipper.Mode.HARDWARE;
+
+    /** Each mode represents a compression algorithm. */
+    public enum Mode {
+        /** QAT deflate mode. */
+        QAT_DEFLATE("QATDEFLATE99", Set.of("qat_deflate")),
```

Comment:
I'm guessing since the mode and codec are different, the data that is compressed via this codec will not be readable if the QAT is disabled in the future. Might have to call this out in the documentation.

Reply:
On systems where QAT is not available, users can set the `index.codec.qatmode` to `auto` and have a software execution.